pkg/scalers/openstack_metrics_scaler.go

```diff
@@ -0,0 +1,368 @@
+	}
+
+	if val, ok := triggerMetadata["metricID"]; ok && val != "" {
+		meta.metricID = val
+	} else {
+		openstackMetricLog.Error(fmt.Errorf("no metric id could be read"), "Error reading metricID")
+		return nil, fmt.Errorf("no metric id was declared")
+	}
+
+	if val, ok := triggerMetadata["aggregationMethod"]; ok && val != "" {
+		meta.aggregationMethod = val
+	} else {
+		openstackMetricLog.Error(fmt.Errorf("no aggregation method could be read"), "Error reading aggregation method")
+		return nil, fmt.Errorf("no aggregation method could be read")
+	}
+
+	if val, ok := triggerMetadata["granularity"]; ok && val != "" {
+		granularity, err := strconv.Atoi(val)
+		if err != nil {
+			openstackMetricLog.Error(err, "Error converting granulality information %s", err.Error)
+			return nil, err
+		}
+		meta.granularity = granularity
+	} else {
+		return nil, fmt.Errorf("no granularity found")
+	}
+
+	if val, ok := triggerMetadata["threshold"]; ok && val != "" {
+		// converts the string to float64 but its value is convertible to float32 without changing
+		_threshold, err := strconv.ParseFloat(val, 32)
+		if err != nil {
+			openstackMetricLog.Error(err, "error parsing openstack metric metadata", "threshold", "threshold")
+			return nil, fmt.Errorf("error parsing openstack metric metadata : %s", err.Error())
+		}
+
+		meta.threshold = _threshold
+	}
+
+	if val, ok := triggerMetadata["timeout"]; ok && val != "" {
+		httpClientTimeout, err := strconv.Atoi(val)
+		if err != nil {
+			return nil, fmt.Errorf("httpClientTimeout parsing error: %s", err.Error())
+		}
+		meta.timeout = httpClientTimeout
+	} else {
+		meta.timeout = metricDefaultHTTPClientTimeout
+	}
+
+	return &meta, nil
+}
+
+func parseOpenstackMetricAuthenticationMetadata(config *ScalerConfig) (openstackMetricAuthenticationMetadata, error) {
+	authMeta := openstackMetricAuthenticationMetadata{}
+	authParams := config.AuthParams
+
+	if val, ok := authParams["authURL"]; ok && val != "" {
+		authMeta.authURL = authParams["authURL"]
+	} else {
+		return authMeta, fmt.Errorf("authURL doesn't exist in the authParams")
+	}
+
+	if val, ok := authParams["userID"]; ok && val != "" {
+		authMeta.userID = val
+
+		if val, ok := authParams["password"]; ok && val != "" {
+			authMeta.password = val
+		} else {
+			return authMeta, fmt.Errorf("password doesn't exist in the authParams")
+		}
+	} else if val, ok := authParams["appCredentialSecret"]; ok && val != "" {
+		authMeta.appCredentialSecretID = val
+	} else {
+		return authMeta, fmt.Errorf("neither userID or appCredentialSecretID exist in the authParams")
+	}
+
+	return authMeta, nil
+}
+
+func (a *openstackMetricScaler) GetMetricSpecForScaling() []v2beta2.MetricSpec {
+	targetMetricVal := resource.NewQuantity(int64(a.metadata.threshold), resource.DecimalSI)
+	externalMetric := &v2beta2.ExternalMetricSource{
+		Metric: v2beta2.MetricIdentifier{
+			Name: kedautil.NormalizeString(fmt.Sprintf("openstack-metric-%s", a.metadata.aggregationMethod)),
```

Comment:
The generated metricName have to be unique, so in case there are multiple OpenStack Metric Scalers defined in one ScaledObject, we are able to distingush between them.

It seems to me that the current approach based on `aggregationMethod` is not enough.

Reply:
Right, well, I guess I will append to it the metric-id and the threshold too so. Thanks for pointing it.

Reply:
Sounds good!